include/nbl/asset/IDescriptorSetLayout.h

```diff
@@ -58,310 +32,356 @@ enum E_DESCRIPTOR_TYPE : uint8_t
-			core::vector<SBindingRedirectBuildInfo> buildInfo_descriptors[asset::EDT_COUNT];
-			core::vector<SBindingRedirectBuildInfo> buildInfo_samplers;
+		friend class ICPUDescriptorSetLayout;
+		inline CBindingRedirect clone() const
+		{
+			CBindingRedirect result;
+			result.m_count = m_count;
 
-			for (auto b = _begin; b != _end; ++b)
+			if (result.m_count > 0)
 			{
-				buildInfo_descriptors[b->type].emplace_back(b->binding, b->count);
-				if (b->type == EDT_COMBINED_IMAGE_SAMPLER && b->samplers == nullptr)
-				{
-					buildInfo_samplers.emplace_back(b->binding, b->count);
-					m_mutableSamplerCount += b->count;
-				}
+				result.init();
+				memcpy(result.m_data.get(), m_data.get(), getRequiredMemorySize());
 			}
 
-			for (auto type = 0u; type < asset::EDT_COUNT; ++type)
-				m_descriptorRedirects[type] = SBindingRedirect(buildInfo_descriptors[type].size());
-			m_samplerRedirects = SBindingRedirect(buildInfo_samplers.size());
+			return result;
+		}
+
+		inline size_t conservativeSizeEstimate() const { return getRequiredMemorySize() + sizeof(*this); }
 
-			auto buildRedirect = [](SBindingRedirect& redirect, core::vector<SBindingRedirectBuildInfo>& info)
-			{
-				std::sort(info.begin(), info.end());
+		uint32_t m_count = 0u;
 
-				for (size_t i = 0u; i < info.size(); ++i)
-				{
-					redirect.bindings[i] = info[i].binding;
-					redirect.offsets[i] = info[i].count;
-				}
+		binding_number_t* m_bindingNumbers = nullptr;
+		core::bitflag<typename SBinding::E_CREATE_FLAGS>* m_createFlags = nullptr;
+		core::bitflag<IShader::E_SHADER_STAGE>* m_stageFlags = nullptr;
+		storage_offset_t* m_storageOffsets = nullptr;
+
+		std::unique_ptr<uint8_t[]> m_data = nullptr;
+	};
 
-				std::exclusive_scan(redirect.offsets, redirect.offsets + info.size(), redirect.offsets, 0u);
-			};
+	// utility functions
+	static inline void fillBindingsSameType(SBinding* bindings, uint32_t count, IDescriptor::E_TYPE type, const uint32_t* counts=nullptr, asset::IShader::E_SHADER_STAGE* stageAccessFlags=nullptr)
+	{
+		for (auto i=0u; i<count; i++)
+		{
+			bindings[i].binding = i;
+			bindings[i].type = type;
+			bindings[i].count = counts ? counts[i]:1u;
+			bindings[i].stageFlags = stageAccessFlags ? stageAccessFlags[i]:asset::IShader::ESS_ALL;
+			bindings[i].samplers = nullptr;
+		}
+	}
+
+	bool isIdenticallyDefined(const IDescriptorSetLayout<sampler_type>* _other) const
+	{
+		if (!_other || getTotalBindingCount() != _other->getTotalBindingCount())
+			return false;
+
+		for (uint32_t t = 0u; t < static_cast<uint32_t>(IDescriptor::E_TYPE::ET_COUNT); ++t)
+		{
+			const auto& lhs = m_descriptorRedirects[t];
+			const auto& rhs = _other->m_descriptorRedirects[t];
 
-			for (auto type = 0u; type < asset::EDT_COUNT; ++type)
-				buildRedirect(m_descriptorRedirects[type], buildInfo_descriptors[type]);
-			buildRedirect(m_samplerRedirects, buildInfo_samplers);
+			const auto bindingCount = lhs.getBindingCount();
+			assert(bindingCount==rhs.getBindingCount());
```

Comment:
you cannot assert, this is used to compare any two DS Layouts together, you need to `return false` instead

Reply:
Resolved.